### PR TITLE
Configuration: Auto-generation

### LINKED ISSUE #797: Autoconfiguration option
JSCS should have a cli option called `--configure`. When ran from the root of your project, it will do the following:

1. ask you to specify the location of your javascript files. 
2. run each preset against your codebase.
3. output the number of violations from each preset with a link to the style guide.
4. ask you to choose a preset based on the output of 3.
5. go through the remaining violations, rule by rule. For each violation, it will ask: "Would you like to create an exception for this rule, or fix this issue?"
  1. for each "exception" created, it will nullfy that rule in .jscsrc.
  2. for each "fix this issue", in the future, will automatically fix it, but for now require you to fix it.
6. at the end, it will output a .jscsrc file for you to commit yourself.

Supercedes #341 

<bountysource-plugin>

---
Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/6604741-autoconfiguration-option?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mrjoelkemp**
TODO:

- [x] Fix bug when using auto-config with a directory
- [x] Handle case when user does not supply a path for auto-configure. Show error message
- [x] Readme
- [ ] Extract and test my custom promisify for 100% coverage (Only if still needed after Inquire switch) https://github.com/jscs-dev/node-jscs/issues/999
- [ ] Use https://github.com/SBoudrias/Inquirer.js instead of prompt for better error handling https://github.com/jscs-dev/node-jscs/issues/1000
- [x] Add test for cli with `--auto-configure` used without a string value (empty defaults to boolean true)
- [ ] Add test for an es6 codebase. We may need to have the esnext flag passed through to the checker used in generator https://github.com/jscs-dev/node-jscs/issues/1001
- [ ] Add tests for erroneous input handling https://github.com/jscs-dev/node-jscs/issues/1002
- [ ] Speed up auto-configure tests https://github.com/jscs-dev/node-jscs/issues/1003

---

- [x] Refactor
- [x] Tests
 - Choosing a preset with no errors
 - Choosing a preset with errors
 - Console logging
- [x] Fix commit message
- [ ] Check that prompt works when the terminal is resized. Seemed broken when stretching terminal to fit everything on one line. (Could be fixed with a move to Inquire, but not sure) #1000

Fixes #797

### CODE REVIEW

File: lib/config/generate.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+    return Object.keys(errorNameLUT);
+}
+
+function getPresetChoice(choice, presetNames) {
+    var presetIndex = choice[prompts[0]] - 1;
+    var presetName = presetNames[presetIndex];
+
+    if (!presetName) {
+        console.error('You supplied an invalid preset number');
```
Comment (mrjoelkemp) [inline lib/config/generate.js]: Instead, try to use prompt's built-in validation

File: lib/config/generate.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    prompt.start();
+    prompt.message = '';
+    prompt.delimiter = '';
+
+    return Vow.all(presetNames.map(checkAgainstPreset.bind(null, _path)))
+    .then(function(results) {
+        var errorsCollection = [].concat.apply([], results);
+        var errorCounts = getErrorCounts(errorsCollection);
+
+        outputPresetErrorCounts(presetNames, errorCounts);
```
Comment (mrjoelkemp) [inline lib/config/generate.js]: Only output unique errors to avoid count discrepancy between the table and the number of rules the user is walked through for fixup/exception.

**Comment by markelog** [comment]
I think we can release 1.11 right after it's done.

/cc @mikesherov @mdevils

**Comment by mikesherov** [comment]
Looks good what you have so far. 

One thought is that when no config is found, if you're TTY, go straight to auto configure mode. If you're not TTY, mention the configure option in the error output.

**Comment by mikesherov** [comment]
Second thought is that if you run configure mode without a path, prompt the user for the path. Combined with my previous suggestion, the user can just literally do: `npm install jscs && jscs` from their root and be walked through the setup.

**Comment by indexzero** [comment]
Can't want to try this out. I basically spent week making config files before >_<

File: lib/config/generate.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+/**
+ * Generates a configuration object based for the given path
+ * based on the best fitting preset
+ *
+ * @param  {String} path - The path containing file(s) used to guide the configuration
+ *
+ * @return {Promise} Resolved with the generated, JSCS configuration
+ */
+function generateConfig(path) {
```
Comment (mdevils) [inline lib/config/generate.js]: I think a simple class would be nicer and more obvious here.

**Comment by mdevils** [comment]
Great job, Joel!

File: lib/config/generate.js (hunk 3 above)
Comment (mrjoelkemp) [inline lib/config/generate.js]: Agreed about the clarity. Forcing an instantiation for what should be a function didn't seem correct. This function only looks like a class because the tests need access to internal methods for stubbing. 

I'll convert this to a class though; thanks for reaffirming.

Tests ruin everything :)

File: lib/config/generator.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    }
+};
+
+function getChecker() {
+    var checker = new Checker();
+    checker.registerDefaultRules();
+    return checker;
+}
+
+function normalizePath(filepath, basePath) {
```
Comment (mikesherov) [inline lib/config/generator.js]: This should be extracted to utils too. This same "if relative, then resolve" code is in 4 spots.

File: lib/config/generator.js [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+                    _this._processViolationChoice(userChoice, associatedRuleName);
+                });
+
+                return this._config;
+            });
+        }.bind(this))
+        .then(function flushConfig(config) {
+            fs.writeFileSync(process.cwd() + '/.jscsrc', JSON.stringify(config, null, '\t'));
+            console.log('Generated a .jscsrc configuration file in ' + process.cwd());
+        });
```
Comment (mikesherov) [inline lib/config/generator.js]: Do you put a final `.catch` here for unexpected errors? Or this handled by the code this promise is returned to?

File: lib/config/generator.js [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+        return Vow
+        .all(presetNames.map(this._checkAgainstPreset.bind(this, _path)))
+        .then(function generateErrorsCollection(results) {
+            errorsCollection = [].concat.apply([], results);
+            return errorsCollection;
+        })
+        .then(this._showErrorCounts.bind(this, presetNames))
+        .then(this._getUserPresetChoice.bind(this, prompts[0]))
+        .then(function getPresetIndexFromPrompt(choiceObj) {
```
Comment (mikesherov) [inline lib/config/generator.js]: Combine this and the next two thens. If you want to use functions for clarity, go ahead, but no need for multiple thens, it's just noisy.

File: lib/config/generator.js [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+            var errorPrompts = generateRuleHandlingPrompts(errorList);
+
+            // TODO: Extract method?
+            return this._getUserViolationChoices(errorPrompts)
+            .then(function handleViolatedRules(choices) {
+                errorPrompts.forEach(function(errorPrompt, idx) {
+                    var associatedRuleName = errorList[idx].rule;
+                    var userChoice = choices[errorPrompt];
+
+                    _this._processViolationChoice(userChoice, associatedRuleName);
```
Comment (mikesherov) [inline lib/config/generator.js]: You avoided `_this` everywhere, why not here? Just pass `this` as second arg to forEach.

File: lib/config/generator.js [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        .then(function showViolatedRules(errorList) {
+            var violatedRuleCount = getUniqueErrorNames(errorList).length;
+
+            if (!violatedRuleCount) { return this._config; }
+
+            console.log(_path + ' violates ' + violatedRuleCount + ' rule' + (violatedRuleCount > 1 ? 's' : ''));
+
+            var errorPrompts = generateRuleHandlingPrompts(errorList);
+
+            // TODO: Extract method?
```
Comment (mikesherov) [inline lib/config/generator.js]: Yes.

File: lib/string-checker.js [hunk 9]
```diff
@@ -58,6 +58,22 @@ StringChecker.prototype = {
     },
 
     /**
+     * Gets the registered presets
+     * @return {Object}
+     */
+    getPresets: function() {
```
Comment (mikesherov) [inline lib/string-checker.js]: Needs a test.

File: lib/string-checker.js [hunk 10]
```diff
@@ -58,6 +58,22 @@ StringChecker.prototype = {
...
+        return this._configuration.getPresets();
+    },
+
+    /**
+     * Gets the base path
+     * @return {String}
+     */
+    getBasePath: function() {
```
Comment (mikesherov) [inline lib/string-checker.js]: Needs a test.

File: lib/config/configuration.js [hunk 11]
```diff
@@ -157,6 +157,15 @@ Configuration.prototype.getMaxErrors = function() {
 };
 
 /**
+ * Returns the list of registered presets
+ *
+ * @return {String[]}
+ */
+Configuration.prototype.getPresets = function() {
```
Comment (mikesherov) [inline lib/config/configuration.js]: Needs a test.

File: lib/config/generator.js (hunk 6 above)
Comment (mikesherov) [inline lib/config/generator.js]: Actually, this and next *3* thens.

File: test/string-checker.js [hunk 12]
```diff
@@ -356,4 +356,24 @@ describe('modules/string-checker', function() {
             });
         }
     });
+
+    describe('getRegisteredPresets', function() {
+        it('returns the configuration\'s registered presets', function() {
```
Comment (mikesherov) [inline test/string-checker.js]: This test should actually check that it returns registered presets. The fact that it's a middleman is implementation detail.

File: test/string-checker.js [hunk 13]
```diff
@@ -356,4 +356,24 @@ describe('modules/string-checker', function() {
...
+            var configuration = checker.getConfiguration();
+            var spy = sinon.spy(configuration, 'getRegisteredPresets');
+            checker.getRegisteredPresets();
+            assert.ok(spy.called);
+            spy.restore();
+        });
+    });
+
+    describe('getBasePath', function() {
```
Comment (mikesherov) [inline test/string-checker.js]: Same middleman comment.

File: test/string-checker.js (hunk 12 above)
Comment (mrjoelkemp) [inline test/string-checker.js]: There's already a test for getRegisteredPreset in configuration and this function's sole role is to delegate to that tested method. Doing checks on the returned value is redundant.

What's the minimum checking here that you feel is appropriate?

Comment (mikesherov) [inline test/string-checker.js]: Then either undelegate the method or copy the test from configuration.

Comment (mrjoelkemp) [inline test/string-checker.js]: In all fairness, the other delegated methods in string-checker don't have dedicated tests. I'll revert my hide delegate.

**Comment by mikesherov** [comment]
Wooo! Looks like testing's done?

**Comment by mikesherov** [comment]
Just wanted to give a quick brain dump of follow up tickets: 1. run configure if JSCS ran without a Config and in tty. 2. detect esnext.

File: bin/jscs [hunk 14]
```diff
@@ -15,6 +15,7 @@ program
     .usage('[options] <file ...>')
     .description('A code style linter for programmatically enforcing your style guide.')
     .option('-c, --config [path]', 'configuration file path')
+    .option('--configure [path]', 'auto-generate a JSCS configuration file')
```
Comment (markelog) [inline bin/jscs]: `--configure` and `--config`, `-c` are very similar. Can we try another name for it?

File: lib/config/generator.js [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+prompt.delimiter = '';
+prompt.start();
+
+/**
+ * Script that walks the user through the autoconfig flow
+ * @type {String[]}
+ */
+var prompts = [
+    {
+        name: 'Please choose a preset number:'.green,
```
Comment (markelog) [inline lib/config/generator.js]: I guess `colors` is used implicitly, i would add it explicitly

File: lib/config/generator.js [hunk 16]
```diff
@@ -0,0 +1,210 @@
+var path = require('path');
+var Vow = require('vow');
+var vowNode = require('vow-node');
+var Table = require('cli-table');
+var prompt = require('prompt');
+var fs = require('fs');
```
Comment (markelog) [inline lib/config/generator.js]: Nit-pick, could we change sequence of those definition like we do in other modules?

Like this:

```js

// Default node modules
var path = require('path');
var fs = require('fs');

// Deps
var Vow = require('vow');
var vowNode = require('vow-node');
var Table = require('cli-table');
var prompt = require('prompt');

// Our stuff
var Checker = require('../checker');
var utils = require('../utils');
```

File: bin/jscs (hunk 14 above)
Comment (zxqfox) [inline bin/jscs]: `--auto-config` or `--auto-generate`?

File: lib/config/generator.js [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+        pattern: /[0-9]*/
+    },
+    {
+        name: 'Create an (e)xception for this rule, or (f)ix the errors yourself?',
+        require: true,
+        pattern: /(e|f)/
+    }
+];
+
+function Generator() {
```
Comment (markelog) [inline lib/config/generator.js]: Could we add JSDoc to the constructor?

File: lib/config/generator.js [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+    this._config = {};
+}
+
+Generator.prototype = {
```
Comment (markelog) [inline lib/config/generator.js]: Nit-pick: it seems, constructor ref is now lost?

File: lib/config/generator.js [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+var Checker = require('../checker');
+var utils = require('../utils');
+
+prompt.message = '';
+prompt.delimiter = '';
+prompt.start();
+
+/**
+ * Script that walks the user through the autoconfig flow
+ * @type {String[]}
```
Comment (markelog) [inline lib/config/generator.js]: @private?

File: lib/config/generator.js [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+     * @private
+     * @param  {Object[]} errorPrompts
+     * @return {Promise}
+     */
+    _getUserViolationChoices: function(errorPrompts) {
+        return this._showPrompt(errorPrompts);
+    },
+
+    /** @private */
+    _showPrompt: vowNode.promisify(prompt.get.bind(prompt)),
```
Comment (markelog) [inline lib/config/generator.js]: It seems as overkill, to add a whole module just for one function

File: lib/config/generator.js [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+    _checkAgainstPreset: function(path, presetName) {
+        var checker = getChecker();
+
+        checker.configure({preset: presetName});
+
+        return checker.checkPath(path);
+    }
+};
+
+function getChecker() {
```
Comment (markelog) [inline lib/config/generator.js]: Nit-pick: missing JSDoc for those helpers?

File: lib/utils.js [hunk 22]
```diff
@@ -158,6 +160,47 @@ exports.trimUnderscores = function(name) {
...
+    return start === './' || start === '..';
+};
+
+/**
+ * Shallow extend
+ * @param  {Object} obj1 - Object to be extended
+ * @param  {Object} obj2
+ * @return {Object} A reference to obj1
+ */
+exports.extend = function(obj1, obj2) {
```
Comment (markelog) [inline lib/utils.js]: Maybe we just should a `lodash` dep?

File: lib/utils.js [hunk 23]
```diff
@@ -158,6 +160,47 @@ exports.trimUnderscores = function(name) {
...
+
+/**
+ * Resolves a relative filepath against the supplied base path
+ * or just returns the filepath if not relative
+ *
+ * @param  {String} filepath
+ * @param  {String} basePath
+ * @return {String}
+ */
+exports.normalizePath = function(filepath, basePath) {
```
Comment (markelog) [inline lib/utils.js]: Should it be in `utils` if it used in one module? In some projects `utils` quickly becomes a collection of weird functions, could we avoid that?

File: lib/utils.js (hunk 22 above)
Comment (zxqfox) [inline lib/utils.js]: It weights much more, even if https://www.npmjs.com/package/lodash.assign.

File: lib/utils.js [hunk 24]
```diff
@@ -158,6 +160,47 @@ exports.trimUnderscores = function(name) {
 };
 
 /**
+ * Whether or not the given path is relative
+ * @param  {String}  path
+ * @return {Boolean}
+ */
+exports.isRelativePath = function(path) {
```
Comment (markelog) [inline lib/utils.js]: How about 

```js
path.resolve( p ) !== path.normalize( p )
```

File: lib/config/generator.js [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+
+/**
+ * Script that walks the user through the autoconfig flow
+ * @type {String[]}
+ */
+var prompts = [
+    {
+        name: 'Please choose a preset number:'.green,
+        require: true,
+        pattern: /[0-9]*/
```
Comment (markelog) [inline lib/config/generator.js]:
```js
/[0-9]*/.test("a")
true
```

I expect something like:

```js
/\d+/
```

But we have only 8 presets, need to limit input to that number or handle it ourselves.

Comment (markelog) [inline lib/config/generator.js]: I would even use https://github.com/SBoudrias/Inquirer.js here, its more popular, and has more features and flexibility, like https://github.com/SBoudrias/Inquirer.js#user-content-prompts-type would work much better here

**Comment by markelog** [comment]
1. For `jscs --configure` it gives
```
/Users/arkel/Workspace/jscs/lib/utils.js:169
    var start = path.substring(0, 2);
                     ^
TypeError: Object true has no method 'substring'
    at Object.exports.isRelativePath (/Users/arkel/Workspace/jscs/lib/utils.js:169:22)
    at Object.exports.normalizePath (/Users/arkel/Workspace/jscs/lib/utils.js:196:14)
    at Object.Generator.generate (/Users/arkel/Workspace/jscs/lib/config/generator.js:47:27)
    at module.exports (/Users/arkel/Workspace/jscs/lib/cli.js:104:10)
    at Object.<anonymous> (/Users/arkel/Workspace/jscs/bin/jscs:36:1)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
```

It seems `No input files specified.` should we displayed instead.

2. Did `node bin/jscs lib test bin publis` from https://github.com/jscs-dev/node-jscs/blob/ccff2178d35c785190eb5f62a75149a4ae6dd5a6/package.json#L112

Get the same error above, whereas `node bin/jscs lib test bin publish -c .jscsrc` works, looks like a blocker

**Comment by markelog** [comment]
I guess comments above is enough for now

File: lib/utils.js (hunk 22 above)
Comment (markelog) [inline lib/utils.js]: Yeah, but it feels like we could use it in a lot of places

File: lib/config/generator.js (hunk 25 above)
Comment (zxqfox) [inline lib/config/generator.js]: I think we should allow custom presets somehow. It will be useful feature if we allow to reconfigure already existed `.jscsrc` according to new code if rules were not specified in `.jscsrc` but were in preset. Something like auto-nullifier?

Comment (markelog) [inline lib/config/generator.js]: > I think we should allow custom presets somehow.

You mean from the plugin?

> Something like auto-nullifier?

Hm, not sure what you asking, "auto-nullifier" already exist with these changes

Comment (zxqfox) [inline lib/config/generator.js]: > You mean from the plugin?

Yup. I'm not sure if we can look in `node_modules` or `.env.NODE_PATH` for plugins (how to differentiate them from others, by `jscs-` prefix?), but it would be great if we could load `.jscsrc` and read `plugins` or `preset` options, or just provide `--plugins` option to cli to set them explicitly.

Another idea is to have `~/.jscsrc` file and use it as default (with plugins, presets, etc.).

Comment (markelog) [inline lib/config/generator.js]: This question could be deferred, in fact, i would like to do https://github.com/jscs-dev/node-jscs/issues/109 (if that something we're willing to do).

But it seems identification through `jscs-` is our best bet.

**Comment by markelog** [comment]
I can finish this up if @mrjoelkemp is busy, but would love some help with writing the docs.

**Comment by mrjoelkemp** [comment]
@markelog thanks for the offer, but i'll have time tonight and tomorrow night to address some of the comments. Most importantly, i'll fix the blocker (with a test) and write documentation.

**Comment by markelog** [comment]
Awesome!

Will you check out inquirer module?

**Comment by mrjoelkemp** [comment]
It doesn't seem like it'll be a lot of work to switch it to inquirer. I'll
consider it; however, my priority is to unblock the release of 1.11 and do
follow up changes/bugfixes shortly after.

**Comment by markelog** [comment]
Sounds good

File: bin/jscs (hunk 14 above)
Comment (mrjoelkemp) [inline bin/jscs]: Opting for `--auto-configure`

File: lib/config/generator.js (hunk 20 above)
Comment (mikesherov) [inline lib/config/generator.js]: Agreed. Simple enough to promisify it yourself manually.

File: lib/utils.js (hunk 22 above)
Comment (mrjoelkemp) [inline lib/utils.js]: Sure. Opting for lodash.assign.

File: lib/utils.js (hunk 23 above)
Comment (mrjoelkemp) [inline lib/utils.js]: It's used in generator and in configuration. The only suitable location aside from utils would be `checker` if we want to avoid a PathHelper.

File: lib/utils.js (hunk 24 above)
Comment (mrjoelkemp) [inline lib/utils.js]: I think the existing solution is much clearer and more performant. Any edge cases that your method accounts for?

File: lib/config/generator.js (hunk 20 above)
Comment (mrjoelkemp) [inline lib/config/generator.js]: A minor annoyance that I have is rolling my own functions and writing tests for them only to have it suggested that I scrap that effort and use a 3rd party solution – or the reverse (using 3rd party and asked to roll my own). 

Is a way to avoid the deletion of work to keep functions in closed scope and during a review, have other maintainers suggest the extraction and testing of the function if they know of other places where it could be applied?

Comment (zxqfox) [inline lib/config/generator.js]: If you look at this module code: https://github.com/dfilatov/vow-node/blob/master/lib/vow-node.js there are just 2 functions. It's a kind of micromodule as well as lodash.assign.

I'm not against it as soon as we accept lodash.assign. ;-)

File: lib/config/generator.js (hunk 21 above)
Comment (zxqfox) [inline lib/config/generator.js]: I have a suggestion. Can we use `jscs-jsdoc` plugin a little bit later and fix all docs?

Sure it would be great if these new functions will have some description to decrease well known bus factor.

File: lib/config/generator.js (hunk 5 above)
Comment (mrjoelkemp) [inline lib/config/generator.js]: That's handled by the caller: cli.

File: lib/config/generator.js (hunk 21 above)
Comment (mrjoelkemp) [inline lib/config/generator.js]: I think that's a great idea. We should add an issue to do that in 1.12 or 1.later depending on priorities.

File: lib/utils.js (hunk 24 above)
Comment (zxqfox) [inline lib/utils.js]: Windows maybe? Not sure.

**Comment by mrjoelkemp** [comment]
@markelog Addressed a bunch of your issues. This PR is still not ready yet; the outstanding todo items are listed in the PR's description. Will ping you when this is ready for a re-review.

**Comment by zxqfox** [comment]
@mrjoelkemp You've decided to not use vow-node? ;-) That's strange since lodash.assign is here.

**Comment by mrjoelkemp** [comment]
@zxqfox Yeah :) I do agree that lodash.assign has the potential to be more valuable to the codebase than vow-node. I may end up not even using my custom promisify with the switch to Inquirer.js.

**Comment by zxqfox** [comment]
:+1: That's ok. I just still remember that issue where @dfilatov decided to split promisify to separated module to make core codebase lighter for browser usage ;-)

File: lib/cli.js [hunk 26]
```diff
@@ -62,7 +63,7 @@ module.exports = function(program) {
         return returnArgs;
     }
 
-    if (!args.length && process.stdin.isTTY) {
+    if (!args.length && process.stdin.isTTY && !program.autoConfigure) {
```
Comment (mikesherov) [inline lib/cli.js]: as @markelog pointed out, this seems wrong. You still need to call autoconfigure on a set of files.

**Comment by mrjoelkemp** [comment]
@markelog I fixed the bug where using `--auto-configure` without a path shows the "No input files specified" message. This still needs a test (as the todo list suggests).

In terms of usage (as the readme states), this was built to be used like `jscs --auto-configure path` where path is a file or directory. I hadn't considered supplying multiple paths. I now see the utility in that for legacy codebases, but is this a blocker? If we change the format for the auto-configure cli option, is that considered "breaking"?

I'm concerned with continuing to block 1.11. As it stands, auto-configure works for a subset of cases. It does not properly handle erroneous input, does not use Inquire,js, does not turn on automatically the way Mike suggested, doesn't work for ES6 codebases, and only supports single input paths (file or directory, but not multiple args).

We should either ship 1.11 without this (realistically, it'll be at *least* another week to get some of the missing things in) or are we fine with shipping a first draft of this and collectively iterating on it thereafter? cc @mikesherov @mdevils @zxqfox

**Comment by zxqfox** [comment]
I would do it because anybody wishes this a lot and it doesn't break anything. But it should be noticed as experimental. And I hope it does not break any current cli usage.

If we strongly need here some patches I can help. Personally I'm missing ES6 :-(

Great job, Joel! :tada:

**Comment by markelog** [comment]
I'd say better ship early, there probably more issues with it that we didn't consider yet. If we would try to make everything perfect, we would never release anything.

> But it should be noticed as experimental

+1

@mrjoelkemp 
> does not turn on automatically

This is 2.0 stuff https://github.com/jscs-dev/node-jscs/issues/986 it would be great to have open tickets for other issues you mentioned.

**Comment by mikesherov** [comment]
Fuck it, we'll do it live!

@mrjoelkemp, please file a new issue with remaining items, and let's land this thing!

**Comment by mrjoelkemp** [comment]
Will do. I'll merge today after one last pass through.
